ipa-core/src/protocol/context/dzkp_field.rs

```diff
@@ -12,17 +25,332 @@ pub trait DZKPCompatibleField: Field {
+    // e = ab⊕cd⊕ f
+    fn convert_prover<'a>(
+        x_left: &'a BitArray32,
+        x_right: &'a BitArray32,
+        y_left: &'a BitArray32,
+        y_right: &'a BitArray32,
+        prss_right: &'a BitArray32,
+    ) -> impl Iterator<Item = UVPolynomials<Fp61BitPrime>> {
+        x_left
+            .chunks(RECURSION_CHUNK_SIZE)
+            .zip(y_right.chunks(RECURSION_CHUNK_SIZE))
+            .zip(y_left.chunks(RECURSION_CHUNK_SIZE))
+            .zip(x_right.chunks(RECURSION_CHUNK_SIZE))
+            .zip(prss_right.chunks(RECURSION_CHUNK_SIZE))
+            .map(|((((a, b), c), d), f)| {
+                // e = ab⊕cd⊕ f = x_left * y_right ⊕ y_left * x_right ⊕ prss_right
+                let e: BitArray<[u8; RECURSION_CHUNK_SIZE >> 3], Lsb0> = (BitArray::try_from(a)
+                    .unwrap()
+                    & BitArray::<[u8; RECURSION_CHUNK_SIZE >> 3], Lsb0>::try_from(b).unwrap())
+                    ^ (BitArray::<[u8; RECURSION_CHUNK_SIZE >> 3], Lsb0>::try_from(c).unwrap()
+                        & BitArray::<[u8; RECURSION_CHUNK_SIZE >> 3], Lsb0>::try_from(d).unwrap())
+                    ^ BitArray::<[u8; RECURSION_CHUNK_SIZE >> 3], Lsb0>::try_from(f).unwrap();
+                // precompute ac = a & c
+                let ac: BitArray<[u8; RECURSION_CHUNK_SIZE >> 3], Lsb0> = BitArray::try_from(a)
+                    .unwrap()
+                    & BitArray::<[u8; RECURSION_CHUNK_SIZE >> 3], Lsb0>::try_from(c).unwrap();
+                // precompute bd = b & c
+                let bd: BitArray<[u8; RECURSION_CHUNK_SIZE >> 3], Lsb0> = BitArray::try_from(b)
+                    .unwrap()
+                    & BitArray::<[u8; RECURSION_CHUNK_SIZE >> 3], Lsb0>::try_from(d).unwrap();
+                (
+                    // g polynomial
+                    a.iter()
+                        .zip(c.iter())
+                        .zip(e.iter())
+                        .zip(ac.iter())
+                        .flat_map(|(((a, c), e), ac)| {
+                            let one_minus_two_e = Fp61BitPrime::ONE
+                                + Fp61BitPrime::MINUS_TWO * Fp61BitPrime::from_bit(*e);
+                            [
+                                // g1=-2ac(1-2e),
+                                Fp61BitPrime::MINUS_TWO
+                                    * Fp61BitPrime::from_bit(*ac)
+                                    * one_minus_two_e,
+                                // g2=c(1-2e),
+                                Fp61BitPrime::from_bit(*c) * one_minus_two_e,
+                                // g3=a(1-2e),
+                                Fp61BitPrime::from_bit(*a) * one_minus_two_e,
+                                // g4=-1/2(1-2e),
+                                Fp61BitPrime::MINUS_ONE_HALF * one_minus_two_e,
+                            ]
+                        })
+                        .collect::<GenericArray<Fp61BitPrime, InitialRecursionFactor>>(),
+                    // h polynomial
```

Comment:
I don't understand how `InitialRecursionFactor` came into play here. It seems to me like you're saying the output of this flat_map should have length exactly equal to 32. How do we know that's going to be the case? Each multiplication tuple is generating 8 values (g1, g2, g3, g4, h1, h2, h3, h4). We are Iterating through all of `a` here which is a whatever we get by iterating through `x_left.chunks(RECURSION_CHUNK_SIZE)`. So it has something to do with 8... I'm just lost.

Reply:
chunks collects 8 bits, for each of the 8 bits I generate a,c,e,ac which define 4 field elements each. So I obtain 32 in total that I collect into a GenericArray.